pkg/kv/cosmosdb/store.go

```diff
@@ -0,0 +1,324 @@
+			return kv.ErrPredicateFailed
+		}
+		return err
+	} else if valuePredicate == kv.PrecondConditionalExists {
+		patch := azcosmos.PatchOperations{}
+		patch.AppendReplace("/value", item.Value)
+		_, err = s.containerClient.PatchItem(
+			ctx,
+			pk,
+			item.ID,
+			patch,
+			&itemOptions,
+		)
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == http.StatusNotFound {
+			return kv.ErrPredicateFailed
+		}
+		return err
+	}
+
+	etag := azcore.ETag(valuePredicate.([]byte))
+	itemOptions.IfMatchEtag = &etag
+	_, err = s.containerClient.UpsertItem(ctx, pk, b, &itemOptions)
+	var respErr *azcore.ResponseError
+	if errors.As(err, &respErr) && respErr.StatusCode == http.StatusPreconditionFailed {
+		return kv.ErrPredicateFailed
+	}
+	return err
+}
+
+func (s *Store) Delete(ctx context.Context, partitionKey, key []byte) error {
+	if len(partitionKey) == 0 {
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	pk := azcosmos.NewPartitionKeyString(s.encoding.EncodeToString(partitionKey))
+
+	_, err := s.containerClient.DeleteItem(ctx, pk, s.encoding.EncodeToString(key), nil)
+	var respErr *azcore.ResponseError
+	if errors.As(err, &respErr) && respErr.StatusCode != http.StatusNotFound {
+		return err
+	}
+	return nil
+}
+
+func (s *Store) Scan(ctx context.Context, partitionKey []byte, options kv.ScanOptions) (kv.EntriesIterator, error) {
+	if len(partitionKey) == 0 {
+		return nil, kv.ErrMissingPartitionKey
+	}
+
+	pk := azcosmos.NewPartitionKeyString(s.encoding.EncodeToString(partitionKey))
+
+	queryPager := s.containerClient.NewQueryItemsPager("select * from c where c.id >= @start order by c.id ", pk, &azcosmos.QueryOptions{
+		ConsistencyLevel: azcosmos.ConsistencyLevelBoundedStaleness.ToPtr(),
+		PageSizeHint:     int32(options.BatchSize),
+		QueryParameters: []azcosmos.QueryParameter{{
+			Name:  "@start",
+			Value: s.encoding.EncodeToString(options.KeyStart),
+		}},
+	})
+	currPage, err := queryPager.NextPage(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EntriesIterator{
+		queryPager: queryPager,
+		currPage:   currPage,
+		queryCtx:   ctx,
+		encoding:   s.encoding,
+	}, nil
+}
+
+func (s *Store) Close() {
+}
+
+type EntriesIterator struct {
+	entry        *kv.Entry
+	err          error
+	currEntryIdx int
+	queryPager   *runtime.Pager[azcosmos.QueryItemsResponse]
+	queryCtx     context.Context
+	currPage     azcosmos.QueryItemsResponse
+	encoding     *base32.Encoding
+}
+
+func (e *EntriesIterator) Next() bool {
+	if e.err != nil {
+		return false
+	}
+
+	if e.currEntryIdx >= len(e.currPage.Items) {
+		if !e.queryPager.More() {
+			return false
+		}
+		var err error
+		e.currPage, err = e.queryPager.NextPage(e.queryCtx)
+		if err != nil {
+			e.err = err
+			return false
+		}
+		e.currEntryIdx = 0
+	}
+
+	var itemResponseBody Document
+	err := json.Unmarshal(e.currPage.Items[e.currEntryIdx], &itemResponseBody)
```

Comment:
Can currPage (or getting a empty results from a query) be without items?
If need to verify that `Items` include entry at index before we try to unmarshal.

Reply:
It can be empty. But then the previous if block takes care of that, isn't it?